### DISCUSSION TITLE
Please guide me where the KeepAliveTimeout server check is implemented

### DISCUSSION THREAD

**Discussion by Antiarchitect** [ntex-rs/ntex-mqtt#91] [Q&A]
According to MQTT spec:
It is the responsibility of the Client to ensure that the interval between Control Packets being sent does not exceed the Keep Alive value. In the absence of sending any other Control Packets, the Client MUST send a PINGREQ Packet [MQTT-3.1.2-23].

If the Keep Alive value is non-zero and the Server does not receive a Control Packet from the Client within one and a half times the Keep Alive time period, it MUST disconnect the Network Connection to the Client as if the network had failed [MQTT-3.1.2-24].

I'm interested in `one and a half time the Keep Alive time period` part. So for 30 seconds by default the server should wait for 45 seconds. Is that actually so?

**Comment by fafhrd91** [comment]
server starts keep-alive [here](https://github.com/ntex-rs/ntex-mqtt/blob/655db48e798eb97d8e9702dca0b56f992b245705/src/v3/server.rs#L512) you can override keep-alive period in [handshake service](https://github.com/ntex-rs/ntex-mqtt/blob/655db48e798eb97d8e9702dca0b56f992b245705/src/v3/handshake.rs#L127)

**Comment by Antiarchitect** [comment]
I see it is implemented here:
https://github.com/ntex-rs/ntex/blob/master/ntex-io/src/dispatcher.rs#L126-L139
Is this implementation correct paying attention to `[MQTT-3.1.2-24]`?

**Comment by fafhrd91** [comment]
ntex-io has nothing to do with mqtt, you have to provide right timeout

**Comment by Antiarchitect** [comment]
I've read the explanation here: https://www.hivemq.com/blog/mqtt-essentials-part-10-alive-client-take-over/#:~:text=It%20is%20the%20responsibility%20of%20the%20MQTT%20client%20to%20set,to%20its%20current%20signal%20strength.
And now understand that setting Keep Alive is the client responsibility. 
Do I understand correctly that in order to conform the specs I need to read the client's CONNECT packet's keepalive value (https://docs.rs/ntex-mqtt/latest/ntex_mqtt/v3/codec/struct.Connect.html), use it if it exists or provide some default, multiply it on 1.5, and set this value (45 in case of 30 seconds from the client) in `idle_timeout` function for the handshake ack? Will HandshakeAck idle_timeout set the actual timeout the client will be disconnected on server side?

**Comment by fafhrd91** [comment]
HandshakeAck::idle_timeout represents timeout fir server

**Comment by Antiarchitect** [comment]
Thank you, @fafhrd91, it seems it becomes clear to me.
